typing/printtyp.ml

```diff
@@ -1903,36 +1900,75 @@ type printing_status =
       type error.
   *)
 
-let printing_status  = function
-  | Trace.(Diff { got=t1, t1'; expected=t2, t2'}) ->
-      if  is_constr_row ~allow_ident:true t1'
-       || is_constr_row ~allow_ident:true t2'
-      then Discard
-      else if same_path t1 t1' && same_path t2 t2' then Optional_refinement
-      else Keep
+let diff_printing_status { Errortrace.got=t1, t1'; expected=t2, t2'} =
+  if  is_constr_row ~allow_ident:true t1'
+   || is_constr_row ~allow_ident:true t2'
+  then Discard
+  else if same_path t1 t1' && same_path t2 t2' then Optional_refinement
+  else Keep
+
+type 'variety trace_format = {
+  incompatibility_phrase : string;
+  constraint_escape_status : printing_status;
+  drop_from_trace : 'a. (Types.type_expr * 'a, 'variety) Errortrace.elt -> bool;
+  explain_contextless_escaped_field_mismatch : bool
+}
+
+(* Each of these is used 2/3 times; none are changed in [unification], two are changed in
+   [equality], and the other two in [moregen] *)
+let default_trace_format = {
+  incompatibility_phrase = "is not compatible with type";
+  constraint_escape_status = Discard;
+  drop_from_trace = (fun _ -> false);
+  explain_contextless_escaped_field_mismatch = true
+}
+
+let unification = default_trace_format
+
+let equality = {default_trace_format with
+  incompatibility_phrase = "is not equal to type";
+  drop_from_trace = (function
+    | Diff {got = ({desc = Tpoly _}, _); expected = ({desc = Tpoly _}, _)} -> true
+    | _ -> false);
```

Comment:
Dropping the differences with an universal quantification seems not that useful. Consider:
```ocaml
module M: sig
  class type ['b] t = object
    method m: 'a. 'a -> 'b
  end
end = struct
  class type ['a] t = object
    method m: 'b. 'a -> 'b
  end
end
```
after dropping the `Tpoly` diffs, the error message becomes 

>```
>       Class type declarations do not match:
>         class type ['a] t = object method m : 'a -> 'b end
>       does not match
>         class type ['b] t = object method m : 'a -> 'b end
>       The method m has type 'a -> 'b but is expected to have type 'a0 -> 'b0
>       Type 'a is not compatible with type 'a0 
>```

which seems incomplete compared to
>```
>       The method m has type 'b. 'a -> 'b but is expected to have type
>         'a0. 'a0 -> 'b
>```

Moreover, removing this extra filtering would remove the need for `drop_from_trace` entirely.